fsfreeze.go

```diff
@@ -0,0 +1,150 @@
+func fsfreezeFreeze() (int, error) {
+	frozenFSCount := 0
+	m, err := getMountPoints()
+	if err != nil {
+		return 0, fmt.Errorf("error while getting mountpoints to fsfreeze-freeze: %v", err)
+	}
+
+	for _, mp := range m {
+		fs, err := os.Open(mp.FSFile)
+		if err != nil {
+			return 0, fmt.Errorf("error while opening mount path for fsfreeze-freeze: %v", err)
+		}
+		defer fs.Close()
+
+		if err := ioctl(fs.Fd(), FIFREEZE, 0); err != nil &&
+			err.(*os.SyscallError).Err.(syscall.Errno) != syscall.EOPNOTSUPP &&
+			err.(*os.SyscallError).Err.(syscall.Errno) != syscall.EBUSY {
+			return 0, fmt.Errorf("error while executing fsfreeze-freeze on mount path %q: %v", mp.FSFile, err)
+		}
+
+		frozenFSCount++
+	}
+
+	return frozenFSCount, nil
+}
+
+// fsfreezeThaw unfreezes the frozen filesystems.
+func fsfreezeThaw() (int, error) {
+	thawedFSCount := 0
+
+	m, err := getMountPoints()
+	if err != nil {
+		return 0, fmt.Errorf("error while getting mountpoints to fsfreeze-thaw: %v", err)
+	}
+
+	for _, mp := range m {
+		fs, err := os.Open(mp.FSFile)
+		if err != nil {
+			return 0, fmt.Errorf("error while opening mount path for fsfreeze-thaw: %v", err)
+		}
+		defer fs.Close()
+
+		if err := ioctl(fs.Fd(), FITHAW, 0); err != nil && err.(*os.SyscallError).Err.(syscall.Errno) != syscall.EINVAL {
+			return 0, fmt.Errorf("error while executing fsfreeze-thaw on mount path %q: %v", mp.FSFile, err)
+		}
+
+		thawedFSCount++
+	}
+
+	return thawedFSCount, nil
+}
+
+// getMountPoints returns the freezable filesystems' mountpoints.
+func getMountPoints() ([]MEntry, error) {
+	f, err := os.Open("/proc/self/mounts")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	m := make([]MEntry, 0, 10)
+
+	isExist := func(fsspec string) bool {
+		for _, v := range m {
+			if v.FSSpec == fsspec {
+				return true
+			}
+		}
+		return false
+	}
+
+	scanner := bufio.NewScanner(f)
+
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		// Ignore non relevant filesystems.
+		if len(fields) < 2 || isExist(fields[0]) || fields[0][0] != '/' ||
+			fields[2] == "smbfs" || fields[2] == "cifs" {
+			continue
+		}
+
+		// Ignore /dev/root device, in gokrazy it is a read-only filesystem, no point in freezing it.
+		if strings.HasPrefix(fields[0], "/dev/root") {
+			continue
+		}
+		// Ignore loop devices.
+		if strings.HasPrefix(fields[0], "/dev/loop") {
+			continue
+		}
+		// Ignore dm- devices.
+		st, err := os.Lstat(fields[0])
+		if err != nil {
+			return nil, err
+		}
+		if st.Mode()&os.ModeSymlink != 0 {
+			if s, err := os.Readlink(fields[0]); err != nil {
+				return nil, err
+			} else {
+				fields[0] = filepath.Base(s)
+			}
+		}
+		if strings.HasPrefix(fields[0], "dm-") {
+			continue
+		}
+
+		m = append(m, MEntry{fields[0], fields[1], fields[2]})
+	}
+	if err = scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func ioctl(fd uintptr, request, argp uintptr) (err error) {
```

Comment:
I'd ditch the argp uintptr if it's unused because I don't think this strictly follows Go's pointer liveness rules in https://pkg.go.dev/unsafe#Pointer (see (4))

Reply:
Done